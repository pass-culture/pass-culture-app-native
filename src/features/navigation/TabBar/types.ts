import { ParsedAlgoliaParameters } from 'libs/algolia'

export type TabRouteName = keyof TabParamList

export type TabParamList = {
  Home: { shouldDisplayLoginModal: boolean }
  Search: { parameters: ParsedAlgoliaParameters | null } | undefined
  Bookings: undefined
  Favorites: undefined
  Profile: undefined
  // TODO: ticket 6311: remove the cheat area
  CheatArea: undefined
}
