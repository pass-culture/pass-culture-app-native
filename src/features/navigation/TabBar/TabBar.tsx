import { BottomTabBarOptions, BottomTabBarProps } from '@react-navigation/bottom-tabs'
import React from 'react'
import styled from 'styled-components/native'

import { useAuthContext } from 'features/auth/AuthContext'
import { BicolorBookings } from 'ui/svg/icons/BicolorBookings'
import { BicolorFavorite } from 'ui/svg/icons/BicolorFavorite'
import { BicolorLogo } from 'ui/svg/icons/BicolorLogo'
import { BicolorProfile } from 'ui/svg/icons/BicolorProfile'
import { BicolorSearch } from 'ui/svg/icons/BicolorSearch'
import { BicolorIconInterface } from 'ui/svg/icons/types'
import { Warning } from 'ui/svg/icons/Warning'
import { ColorsEnum, getShadow, getSpacing, Spacer, UniqueColors } from 'ui/theme'

import { useCustomSafeInsets } from '../../../ui/theme/useCustomSafeInsets'

import { TabBarComponent } from './TabBarComponent'
import { TabRouteName } from './types'

const mapRouteToIcon = (
  route: TabRouteName | string
): ((props: BicolorIconInterface) => React.ReactNode) => {
  switch (route) {
    case 'HomeNavigator':
      return BicolorLogo
    case 'Search':
      return BicolorSearch
    case 'Bookings':
      return BicolorBookings
    case 'Favorites':
      return BicolorFavorite
    case 'Profile':
      return BicolorProfile
    // TODO: ticket 6311: remove the cheat area
    case 'CheatArea':
      return Warning
    default:
      return BicolorLogo
  }
}
export const TabBar: React.FC<Pick<
  BottomTabBarProps<BottomTabBarOptions>,
  'state' | 'navigation'
>> = ({ navigation, state }) => {
  const { bottom } = useCustomSafeInsets()
  const authContext = useAuthContext()
  return (
    <MainContainer>
      <RowContainer>
        <Spacer.Row numberOfSpaces={4} />
        {state.routes.map((route, index) => {
          const isSelected = state.index === index
          const routeName = state.routeNames[index]
          const onPress = () => {
            if (isSelected) return

            const shouldNavigateLoggedIn = !route.key.includes('Profile') || authContext.isLoggedIn
            if (shouldNavigateLoggedIn) {
              const event = navigation.emit({
                type: 'tabPress',
                target: route.key,
                canPreventDefault: true,
              })
              if (!event.defaultPrevented) {
                navigation.navigate(route.name)
              }
            } else {
              navigation.navigate('Home', { shouldDisplayLoginModal: true })
            }
          }
          return (
            <TabBarComponent
              key={`key-tab-nav-${index}-${route.key}`}
              isSelected={isSelected}
              bicolorIcon={mapRouteToIcon(routeName)}
              onPress={onPress}
              testID={`tab-${routeName}`}
            />
          )
        })}
        <Spacer.Row numberOfSpaces={4} />
      </RowContainer>
      <SafeAreaPlaceholder safeHeight={bottom} />
    </MainContainer>
  )
}
const RowContainer = styled.View({
  flexDirection: 'row',
})
const SafeAreaPlaceholder = styled.View<{ safeHeight: number }>(({ safeHeight }) => ({
  height: safeHeight,
}))
const MainContainer = styled.View({
  alignItems: 'center',
  border: getSpacing(1 / 4),
  borderColor: ColorsEnum.GREY_LIGHT,
  backgroundColor: UniqueColors.TAB_BAR,
  width: '100%',
  position: 'absolute',
  bottom: 0,
  ...getShadow({
    shadowOffset: {
      width: 0,
      height: getSpacing(1 / 4),
    },
    shadowRadius: getSpacing(1),
    shadowColor: ColorsEnum.BLACK,
    shadowOpacity: 0.2,
  }),
})
