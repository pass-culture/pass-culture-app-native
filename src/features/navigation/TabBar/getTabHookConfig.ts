import { TabRouteName, TabParamList } from './TabStackNavigatorTypes'

export function getTabHookConfig<Screen extends TabRouteName>(
  screen: Screen,
  params?: TabParamList[Screen]
): ['TabNavigator', { screen: Screen; params: TabParamList[Screen] }] {
  return ['TabNavigator', { screen, params }]
}
