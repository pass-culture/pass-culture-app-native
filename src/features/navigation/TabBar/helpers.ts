import { TabRouteName, TabParamList } from './types'

export function getShouldDisplayTab({
  isLoggedIn,
  isBeneficiary,
}: {
  isLoggedIn: boolean
  isBeneficiary: boolean
}) {
  return function (name: TabRouteName): boolean {
    if (name.startsWith('Bookings') && (!isLoggedIn || !isBeneficiary)) {
      return false
    }
    return true
  }
}

export function getTabNavConfig<Screen extends TabRouteName>(
  screen: Screen,
  params?: TabParamList[Screen]
): ['TabNavigator', { screen: Screen; params: TabParamList[Screen] }] {
  return ['TabNavigator', { screen, params }]
}

export const homeNavConfig = getTabNavConfig('Home')
