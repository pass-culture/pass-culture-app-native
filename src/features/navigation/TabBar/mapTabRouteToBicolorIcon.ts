import React from 'react'

import { BicolorFavoriteCount } from 'features/favorites/atoms/BicolorFavoriteCount'
import { BicolorBookings } from 'ui/svg/icons/BicolorBookings'
import { BicolorLogo } from 'ui/svg/icons/BicolorLogo'
import { BicolorProfile } from 'ui/svg/icons/BicolorProfile'
import { BicolorSearch } from 'ui/svg/icons/BicolorSearch'
import { BicolorIconInterface } from 'ui/svg/icons/types'

import { TabRouteName } from './types'

export function mapTabRouteToBicolorIcon(route: TabRouteName): React.FC<BicolorIconInterface> {
  switch (route) {
    case 'Home':
      return BicolorLogo
    case 'Search':
      return BicolorSearch
    case 'Bookings':
      return BicolorBookings
    case 'Favorites':
      return BicolorFavoriteCount
    case 'Profile':
      return BicolorProfile
    default:
      return BicolorLogo
  }
}
