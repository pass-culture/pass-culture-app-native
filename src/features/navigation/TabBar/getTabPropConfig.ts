import { TabParamList, TabRouteName } from 'features/navigation/TabBar/TabStackNavigatorTypes'

export function getTabPropConfig<Screen extends TabRouteName>(
  screen: Screen,
  params?: TabParamList[Screen]
): {
  screen: 'TabNavigator'
  params: {
    screen: 'TabNavigator'
    params: { screen: Screen; params: TabParamList[Screen] }
  }
} {
  return {
    screen: 'TabNavigator',
    params: { screen: 'TabNavigator', params: { screen, params } },
  }
}
