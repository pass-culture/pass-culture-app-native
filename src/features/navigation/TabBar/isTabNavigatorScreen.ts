import { rootStackNavigatorPathConfig } from 'features/navigation/RootNavigator/linking/rootStackNavigatorPathConfig'
import { TabRouteName } from 'features/navigation/TabBar/TabStackNavigatorTypes'

export function isTabNavigatorScreen(screen): screen is TabRouteName {
  const tabRouteNames = Object.keys(rootStackNavigatorPathConfig.TabNavigator.screens)
  return tabRouteNames.includes(screen)
}
