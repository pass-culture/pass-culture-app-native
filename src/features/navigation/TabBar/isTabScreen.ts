import { tabStackNavigatorPathConfig } from 'features/navigation/TabBar/TabStackNavigatorPathConfig'
import { TabRouteName } from 'features/navigation/TabBar/TabStackNavigatorTypes'

export function isTabScreen(screen): screen is TabRouteName {
  const tabRouteNames = Object.keys(tabStackNavigatorPathConfig.TabNavigator.screens)
  return tabRouteNames.includes(screen)
}
