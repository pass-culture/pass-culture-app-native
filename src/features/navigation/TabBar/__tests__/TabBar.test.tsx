import { BottomTabNavigationEventMap } from '@react-navigation/bottom-tabs/lib/typescript/src/types'
import { NavigationHelpers, ParamListBase, TabNavigationState } from '@react-navigation/native'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { TabBar } from '../TabBar'

const state: TabNavigationState<Record<string, Record<string, unknown> | undefined>> = {
  history: [
    { key: 'Favorites-key', type: 'route' },
    { key: 'route2-key', type: 'route' },
    { key: 'Bookings-key', type: 'route' },
    { key: 'Profile-key', type: 'route' },
    { key: 'HomeNavigator-key', type: 'route' },
  ],
  index: 0,
  key: 'tab',
  routeNames: ['HomeNavigator', 'Search', 'Bookings', 'Favorites', 'Profile'],
  routes: [
    {
      key: 'HomeNavigator-key',
      name: 'HomeNavigator',
      params: undefined,
    },
    { key: 'Search-key', name: 'Search', params: undefined },
    { key: 'Bookings-key', name: 'Bookings', params: undefined },
    { key: 'Favorites-key', name: 'Favorites', params: undefined },
    { key: 'Profile-key', name: 'Profile', params: undefined },
  ],
  stale: false,
  type: 'tab',
}
const navigation: NavigationHelpers<ParamListBase, BottomTabNavigationEventMap> = {
  canGoBack: jest.fn(),
  dangerouslyGetParent: jest.fn(),
  dangerouslyGetState: jest.fn(),
  dispatch: jest.fn(),
  // @ts-ignore : ignore type of emit to facilitate testing
  emit: jest.fn(() => ({ defaultPrevented: false })),
  goBack: jest.fn(),
  isFocused: jest.fn(),
  navigate: jest.fn(),
  reset: jest.fn(),
  setParams: jest.fn(),
}

jest.mock('react-native-safe-area-context', () => ({
  useSafeAreaInsets: jest.fn(() => ({ bottom: 10 })),
}))
describe('TabBar', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('renders correctly', () => {
    const tabBar = render(<TabBar state={state} navigation={navigation} />)
    expect(tabBar).toMatchSnapshot()
  })
  it('displays only one selected at a time', async () => {
    const tabBar = render(<TabBar state={state} navigation={navigation} />)
    expect(tabBar.queryAllByTestId(/selector/)).toHaveLength(1)
  })
  it('switches tab when clicked on another tab', () => {
    const tabBar = render(<TabBar state={state} navigation={navigation} />)
    expect(tabBar.queryByTestId('selector-tab-HomeNavigator')).toBeTruthy()
    expect(tabBar.queryByTestId('selector-tab-Search')).toBeFalsy()

    const searchTab = tabBar.getByTestId('tab-Search')
    fireEvent.press(searchTab)

    expect(navigation.emit).toHaveBeenCalled()
    expect(navigation.navigate).toHaveBeenCalledWith('Search')
    tabBar.rerender(<TabBar state={{ ...state, index: 1 }} navigation={navigation} />)

    expect(tabBar.queryByTestId('selector-tab-HomeNavigator')).toBeFalsy()
    expect(tabBar.queryByTestId('selector-tab-Search')).toBeTruthy()
  })
  it('does not reset navigation when clicked on selected tab', () => {
    const tabBar = render(<TabBar state={state} navigation={navigation} />)
    expect(tabBar.queryByTestId('selector-tab-HomeNavigator')).toBeTruthy()

    const homeTab = tabBar.getByTestId('tab-HomeNavigator')
    fireEvent.press(homeTab)

    expect(navigation.emit).not.toHaveBeenCalled()
    expect(navigation.navigate).not.toHaveBeenCalled()
  })
})
