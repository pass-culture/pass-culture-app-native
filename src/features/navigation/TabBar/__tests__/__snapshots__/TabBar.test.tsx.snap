// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fffffff8",
        "borderColor": "#f5f5f5",
        "borderStyle": "solid",
        "borderWidth": 1,
        "bottom": 0,
        "position": "absolute",
        "shadowColor": "#151515",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 4,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      numberOfSpaces={4}
      style={
        Array [
          Object {
            "width": 16,
          },
        ]
      }
    />
    <View
      accessibilityLabel="Home tab"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 64,
          "marginTop": -1,
          "opacity": 1,
        }
      }
      testID="Home tab"
    >
      <View
        height={3}
        testID="Home tab selected"
        width={44}
      >
        <Text>
          Home tab selected-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        height={44}
        width={44}
      >
        <Text>
          undefined-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "height": 3,
            },
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="Search tab"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 64,
          "marginTop": -1,
          "opacity": 1,
        }
      }
      testID="Search tab"
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        height={44}
        width={44}
      >
        <Text>
          undefined-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="Bookings tab"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 64,
          "marginTop": -1,
          "opacity": 1,
        }
      }
      testID="Bookings tab"
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        height={44}
        width={44}
      >
        <Text>
          undefined-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="Favorites tab"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 64,
          "marginTop": -1,
          "opacity": 1,
        }
      }
      testID="Favorites tab"
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        height={44}
        width={44}
      >
        <Text>
          undefined-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="Profile tab"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 64,
          "marginTop": -1,
          "opacity": 1,
        }
      }
      testID="Profile tab"
    >
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
      <View
        height={44}
        width={44}
      >
        <Text>
          undefined-SVG-Mock
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
            },
          ]
        }
      />
    </View>
    <View
      numberOfSpaces={4}
      style={
        Array [
          Object {
            "width": 16,
          },
        ]
      }
    />
  </View>
  <View
    safeHeight={5}
    style={
      Array [
        Object {
          "height": 5,
        },
      ]
    }
  />
</View>
`;
